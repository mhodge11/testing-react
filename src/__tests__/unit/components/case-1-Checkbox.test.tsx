/**
 * @jest-environment jsdom 
 */

import React from 'react'
import { axe } from 'jest-axe'
import { render, fireEvent } from '@testing-library/react'

import Checkbox from '../../../components/Checkbox'

/**
 * This checkbox component renders a checkbox with a label.
 * Since we customized the default checkbox, we want to
 * make sure it still works as a regular checkbox
 * should.
 */
describe('The <Checkbox /> component', () => {
  const defaultCheckboxProps = {
    label: 'TEST_CHECKBOX_LABEL',
    id: 'TEST_CHECKBOX_ID',
    background: '#000',
    checkMarkBackground: '#fff',
    onChange: jest.fn(),
    checked: false,
  }

  const setupCheckbox = (props = defaultCheckboxProps) => render(<Checkbox {...props} />)

  it('Should render the label and checkbox the user will see', () => {
    const { asFragment } = setupCheckbox()
    expect(asFragment()).toMatchSnapshot()
  })

  it('Should make the checkbox accessible by setting the id and htmlFor attributes on label and checkbox', () => {
    const { getByLabelText } = setupCheckbox()
    const checkbox = getByLabelText(defaultCheckboxProps.label)
    expect(checkbox).toBeInTheDocument()
  })

  it('Should call the onChange handler when it is provided', () => {
    const { getByLabelText } = setupCheckbox()
    const checkbox = getByLabelText(defaultCheckboxProps.label)
    fireEvent.click(checkbox)
    expect(defaultCheckboxProps.onChange).toHaveBeenCalled()
  })

  it('Should change state correctly when clicked (checked and unchecked)', () => {
    const { getByLabelText } = setupCheckbox({ ...defaultCheckboxProps, checked: true})
    const checkbox = getByLabelText(defaultCheckboxProps.label)
    expect(checkbox).toBeChecked()
  })

  it('should not fail any accessibility tests', async () => {
    const { container } = setupCheckbox()
    const results = await axe(container)
    expect(results).toHaveNoViolations()
  })
})
