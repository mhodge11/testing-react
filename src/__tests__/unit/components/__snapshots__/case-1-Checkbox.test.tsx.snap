// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The <Checkbox /> component Should render the label and checkbox the user will see 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe"
    data-testid="Checkbox"
  >
    <input
      class="sc-hLBbgP fsZgSt"
      id="TEST_CHECKBOX_ID"
      type="checkbox"
    />
    <label
      class="sc-dkrFOg emJCod"
      for="TEST_CHECKBOX_ID"
    >
      <span
        class="sc-gswNZR jluZTW"
      >
        TEST_CHECKBOX_LABEL
      </span>
    </label>
  </div>
</DocumentFragment>
`;
